qbraid/interface/qbraid_qasm/test/test_qasm_utils.py

```diff
@@ -28,3 +40,227 @@ def test_qasm_depth():
+
+
+def test_convert_to_qasm_3():
+    """test the conversion of qasm 2 to 3"""
+    lib_dir = os.path.dirname(os.path.dirname(__file__)) + "/qasm_lib"
+    gate_def_qasm3 = open(
+        os.path.join(lib_dir, "qelib_qasm3.qasm"), mode="r", encoding="utf-8"
+    ).read()
+
+    # 1. qubit statement conversion
+    test_qregs = """OPENQASM 2.0;
+    include "qelib1.inc";
+    qreg q[1] ;
+    qreg qubits  [10]   ;
+    creg c[1];
+    creg bits   [12]   ;
+    """
+    test_qregs_expected = f"""OPENQASM 3.0;
+    include "stdgates.inc";
+    {gate_def_qasm3}
+    qubit[1] q;
+    qubit[10] qubits;
+    bit[1] c;
+    bit[12] bits;"""
+
+    _check_output(convert_to_qasm_3(test_qregs), test_qregs_expected)
+
+    # 2. Measure statement conversion
+    test_measure = """
+    OPENQASM 2.0;
+    include "qelib1.inc";
+    qreg q[2];
+    creg c[2];
+    measure q->c;
+    measure q[0] -> c[1];
+    """
+
+    test_measure_expected = f"""
+    OPENQASM 3.0;
+    include "stdgates.inc";
+    {gate_def_qasm3}
+    qubit[2] q;
+    bit[2] c;
+    c = measure q;
+    c[1] = measure q[0];
+    """
+    _check_output(convert_to_qasm_3(test_measure), test_measure_expected)
+
+    # 3. Opaque comment conversion
+    test_opaque = """
+    OPENQASM 2.0;
+    include "qelib1.inc";
+    qreg q[2];
+    creg c[2];
+    opaque custom_gate (a,b,c) p,q,r;
+    """
+
+    test_opaque_expected = f"""
+    OPENQASM 3.0;
+    include "stdgates.inc";
+    {gate_def_qasm3}
+    qubit[2] q;
+    bit[2] c;
+    // opaque custom_gate (a,b,c) p,q,r;
+    """
+
+    _check_output(convert_to_qasm_3(test_opaque), test_opaque_expected)
+
+    # 4. std header change
+    test_header = """
+    OPENQASM 2.0;
+    include "qelib1.inc";
+    qreg q[1];
+    """
+
+    test_header_expected = f"""
+    OPENQASM 3.0;
+    include "stdgates.inc";
+    {gate_def_qasm3}
+    qubit[1] q;
+    """
+    _check_output(convert_to_qasm_3(test_header), test_header_expected)
+
+    # 5. Unsupported gate conversion
+    test_unsupported = """
+    OPENQASM 2.0;
+    include "qelib1.inc";
+    qreg q[5];
+    u(1,2,3) q[0];
+    sxdg q[0];
+    csx q[0], q[1];
+    cu1(0.5) q[0], q[1];
+    cu3(1,2,3) q[0], q[1];
+    rzz(0.5) q[0], q[1];
+    rccx q[0], q[1], q[2];
+    rc3x q[0], q[1], q[2], q[3];
+    c3x q[0], q[1], q[2], q[3];
+    c3sqrtx q[0], q[1], q[2], q[3];
+    c4x q[0], q[1], q[2], q[3], q[4];
+    """
+    test_unsupported_expected = f"""
+    OPENQASM 3.0;   
+    include "stdgates.inc";
+    {gate_def_qasm3}
+    qubit[5] q;
+    u(1,2,3) q[0];
+    sxdg q[0];
+    csx q[0], q[1];
+    cu1(0.5) q[0], q[1];
+    cu3(1,2,3) q[0], q[1];
+    rzz(0.5) q[0], q[1];
+    rccx q[0], q[1], q[2];
+    rc3x q[0], q[1], q[2], q[3];
+    c3x q[0], q[1], q[2], q[3];
+    c3sqrtx q[0], q[1], q[2], q[3];
+    c4x q[0], q[1], q[2], q[3], q[4];
+    """
+    _check_output(convert_to_qasm_3(test_unsupported), test_unsupported_expected)
+
+
+def test_random_conversion_to_qasm3():
```

Comment:
For this, when we generate 200 random circuits and throw out ones with unsupported gates in a single test function, we don't actually know how many tests we're running, so going forward if something breaks, it's hard to figure out where the error is or how badly things are breaking.

Instead, let's create a `@pytest.fixture` or equivalent helper function that is called beforehand to generate a list of supported test circuits. In this function, you can use your same code to generate random circuits, check if each contains any unsupported gates, and if not, append to a testing list until it reaches length 50 or 100.

Then, you can pass the list of supported qasm strings / circuits to this test function using `@pytest.mark.parametrize` and run that number of clean, isolated tests 🚀

Reply:
[9af229b](https://github.com/qBraid/qBraid/pull/243/commits/9af229b56139c9bd6fc4e475e69ec6f26a7c7363) should add this